### PR TITLE
Deprecate forward decls

### LINKED ISSUE #976: Allow mutual recursion without forward declaration
Relevant to @bob-carpenter 's comment here: https://github.com/stan-dev/docs/issues/408#issuecomment-928228449. 

We currently allow functions to call themselves without a forward declaration, but mutual recursion would still require at least some subset of the required functions to be declared. We can instead do a first pass through the function block which only adds each function's signature to the symbol table, and then our second pass as normal which checks the bodies etc. This would mean all functions are available to all other functions in an order-independent way.

After this change, function declarations are only really useful for defining functions in other (stan|c++) files. In the long term we may want to move toward this being the only valid use of them for simplicity.

### PR DESCRIPTION
**Pull request by nhuurre**
#### Submission Checklist

- [X] Run unit tests
- Documentation
    - [x] If a user-facing facing change was made, the documentation PR is here: stan-dev/docs#612
    - [ ] OR, no user-facing changes were made

Closes #976 

The backend simply generates a forward declaration for every function (even nonrecursive ones) which also simplifies functor struct generation a bit.
Also needed to do something smarter for potentially cyclic eigen templates (#1228 heuristic has been broken for a while now because simple recursive functions don't actually need forward decls)

And also I've allowed variables to shadow user defined functions and not just stan math functions. The reason is shown here:
```stan
functions {
  void foo() {
    xyz(); // function visible even without a forward declaration - yay
  }
  void bar() {
    real xyz; // name conflicts with the function `xyz` unless shadowing allowed - oh no
  }
  void xyz() {
  }
}
```

## Release notes

Recursive user-defined functions no longer need forward declarations.

## Copyright and Licensing

Copyright holder: Niko Huurre
By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by WardBrian** [comment]
How would this interact with the external C++ functions feature?

I think a lot of this is good - generating forward decls for everything and removing any language from the documentation about the need for them is good. We can also make function declarations completely order-independent with this, if we’re willing to do 2 passes over the function block in typechecking, which I would like to see as part of a change like this. 

That said, I think we probably want to pair this with something like an `@external` annotation (see https://github.com/stan-dev/design-docs/pull/45) for their use with external C++, before any deprecation warning was emitted.

More specific review on the code to follow, those are just my thoughts based on the PR description

**Comment by WardBrian** [comment]
`extern "C++" real foo(real);` is an interesting suggestion. For what it's worth, I think "being similar to C++" isn't a strong reason for doing something in Stan, but historically that has been the case.

The thing I like about `@extern` is the potential to do something like this:

```stan
@extern("foo.hpp")
real foo(real);
```

This annotation does 3 things:
1. Disables the "declared but not defined" error for _just this function_
2. Disables all code-gen for this function (maybe functors could stay around, but IMO we don't want to generate _anything_ since that means the user must write code which uses our templates if e.g. we generate a declaration)
3. Place the _contents_ of foo.hpp where we would have generated any function code. This solves the "namespace issue" that currently makes external C++ a pain outside RStan.

I think an extern keyword could easily do 1 and 2, but I'm not sure how to do 3 with it.

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:283]: I think this concat can happen in the above logic (such that fun_decls and fun_defns always have the same type and can be trivially concatenated)

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:244]: I really appreciate how much more general this is than the prior code. Am I correct that as a side effect, we no longer generate any functor if the function is never used in a higher order function?

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [reply src/stan_math_backend/Transform_Mir.ml:470]: Is callmap a map from name -> direct calls or name -> all calls (including through other functions)?

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:522]: I think I can follow the above logic well enough to see it is correct, but I'd appreciate some comments on the major steps/flow of information.

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:513]:
```suggestion
      let calls = String.Hash_set.create () in
```

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:516]:
```suggestion

```

I think you should either make the type of callmap at the top a `String.Hash_set.t String.Table.t` or make `calls` a `String.Set.t` from the start, rather than doing this intermediate thing.

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:498]: You can make the scope of this `open` tighter:

```suggestion
    let rec rewrite_stmt = 
      let open Stmt.Fixed in
      function
      | {pattern= Pattern.NRFunApp ((UserDefined (name, _) as kind), args); _}
```

**Review by WardBrian — CHANGES_REQUESTED** [review]
Overall I think this is very nice. Most of my comments are stylistic or asking for more clarification

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:281]: I believe the current test failures (`../../ml/stanc3/test/integration/good/overloading/sir_overloaded_odes.hpp:112:8: error: redefinition of ‘struct sir_overloaded_odes_model_namespace::simple_SIR_variadic2_functor__’`) are because of this not properly "collating" functor structs into one struct per name

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:182]: Actually it's the removal of this function that I think broke the overloading tests. We probably need something like it still (I think any version which doesn't have this sort of global variable ends up being O(N^2) because you need to search for overloads for each struct you generate)

File: src/stan_math_backend/Cpp.ml
Comment (nhuurre) [reply src/stan_math_backend/Cpp.ml:314]: Functor structs' `operator()` methods have a body and `_init=true`. And in the other direction, external function forward declarations turn into a forward declaration and a "function definition" that's secretly just a second declaration with `_init=false`. (we should fix that at some point but I don't think it _breaks_ anything for now...)

File: src/stan_math_backend/Lower_functions.ml
Comment (nhuurre) [reply src/stan_math_backend/Lower_functions.ml:180]: The previous docstring called attention to the side effects. This version has no side effects and the fact that it takes a MIR `fun_def` and produces a LIR `fun_def` and structs for HOFs is pretty obvious from the signature. Is there anything particular you want the docstring to say?

File: src/stan_math_backend/Lower_functions.ml
Comment (nhuurre) [reply src/stan_math_backend/Lower_functions.ml:244]: That's correct.

File: src/stan_math_backend/Transform_Mir.ml
Comment (nhuurre) [reply src/stan_math_backend/Transform_Mir.ml:516]: I added the type change to signal that we are done collecting calls from this function and the value becomes immutable. But of course that's not correct--the code _on the next line_ mutates the sets inside `callmap`! I'll change it to `Hash_set` everywhere.

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:274]: If this is testing for equality of the argument lists then isn't it testing that the function is actually the same, rather than testing that it is an overload?

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:294]: This can be falsely triggered by an overload, I think?

```stan
functions {
   void foo();

   void foo(int a) {
    print(a);
   }
}
```

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:531]: Perhaps I'm missing something, but does this code still not have an order-dependence in the callgraph population? I don't see how it would have information "flow upward" if the final function in the functions block implicated prior ones, for example.

Also, not sure if this is relevant, but might as well ask: can you imagine a situation in which a function is recursive through the use of a higher-order function? Could that exhibit this poor eigen behavior?

File: src/frontend/Deprecation_analysis.ml
Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: We also generally mention if the canonicalize flag can fix this, and if there is an expiration on the deprecation.

The latter requires a bit more discussion of intent. I don't think we need to settle what it looks like, but if the goal is that somewhere down the line there is a special syntax for external functions and the current forward decls are completely gone, that's worth discussing

File: src/stan_math_backend/Lower_functions.ml
Comment (nhuurre) [reply src/stan_math_backend/Lower_functions.ml:274]: Yes. This is definitely a case where the function name is just a random word that was on my mind while writing the code. `is_correct_overload` would have been more accurate.

File: src/stan_math_backend/Lower_functions.ml
Comment (nhuurre) [reply src/stan_math_backend/Lower_functions.ml:294]: Oops, didn't think of that.
I think it's going to be easier to filter out the extra decls during AST -> MIR transformation and then not worry about it in the backend.

File: src/stan_math_backend/Transform_Mir.ml
Comment (nhuurre) [reply src/stan_math_backend/Transform_Mir.ml:531]: If this function could implicate a prior one then that prior function must already implicate this function. We eval()'ed all cycles that would go through this function so at the point where `callgraph` updates it's still cycle-free.

Good point about higher-order functions, this fails on master
```stan
functions {
  vector foo(real x, vector s, matrix y) {
    return ode_rk45(foo, [1]', 0.0, {1.0}, y[2:])[1];
  }
}
parameters {
  real x;
  vector[3] s;
  matrix[3,3] y;
}
transformed parameters {
   vector[3] z = foo(x,s,y);
}
```

Comment (WardBrian) [reply src/stan_math_backend/Transform_Mir.ml:531]: Assuming you can fix that in this framework, it would be good to add that as a test.

We should also probably add the small program you have in the original PR text as a regression test

Comment (WardBrian) [reply src/stan_math_backend/Transform_Mir.ml:531]: After staring at it a bit longer, I think I understand: we only need to insert an eval _somewhere_ in the chain to prevent the issue. So if we're looking at functions A, B, C, it doesn't matter if A doesn't know to eval calls to C, since when we get to C it will know it should.

This is actually even better than a solution which does what I thought it was, since it won't insert more than are strictly needed.

File: test/integration/good/code-gen/cpp.expected
Comment (nhuurre) [inline test/integration/good/code-gen/cpp.expected:25309]: This is overly cautious because the code did compile before but note that if the Stan source line were changed from
```stan
    return reduce_sum(s, y_slice, 1, a, b, c, d, e, f, g, h, i, j, k, l, m,
                      n, o, p, q);
```
to
```stan
    return reduce_sum(s, y_slice, 1, a, b, c[2:], d, e, f, g, h, i, j, k, l, m,
                      n, o, p, q);
```
it blows up without the `eval()`s.

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [inline src/analysis_and_optimization/Optimize.ml:473]: [Comment] Does the fact that this file changed but none of the generated C++ in the tests did suggest to you that everything is good, or that we may be under-testing this particular optimization?

File: src/stan_math_backend/Lower_functions.ml
Comment (WardBrian) [inline src/stan_math_backend/Lower_functions.ml:289]: Is this FIXME intended to stay in for the merge (I'm mostly fine if the answer is yes, just checking)

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:478]: Thanks for these comments, I find them very helpful.

A nitpick:
```suggestion
     size O(n^2) and this algorithm is O(n^3) so it's important to include
```

Just to keep our file encoding simpler I think we should stick to ASCII

**Review by WardBrian — COMMENTED** [review]
I suspect that this + the unresolved comments above are probably the last substantial review this needs before approval.

Would you mind adding a test which shows function definitions being order-free (the snippet in the original PR text would suffice)? Besides that, I would also like for the PR to stan-dev/docs be written before merging this, since this requires some rather noticeable changes to the user-facing docs.

File: src/stan_math_backend/Lower_functions.ml
Comment (nhuurre) [reply src/stan_math_backend/Lower_functions.ml:289]: Yes. It only affects external functions that are used directly in higher-order functions. You can't use math library functions like that either so I think this can wait until we have better support for external functions. It's not too difficult to fix once external functions are marked in some way in the MIR.

File: src/analysis_and_optimization/Optimize.ml
Comment (WardBrian) [reply src/analysis_and_optimization/Optimize.ml:473]: Mind adding that example into `test/integration/good/compiler-optimizations`?

File: test/integration/bad/functions-bad18.stan
Comment (nhuurre) [reply test/integration/bad/functions-bad18.stan:11]: That math library "contextual name shadowing" doesn't make much sense to me and I couldn't find if it's documented anywhere.

Comment (WardBrian) [reply test/integration/bad/functions-bad18.stan:11]: I'm not sure why it was done this way, but my understanding is:

1. It was desirable to let users name variables the same as built in functions, because names like `beta` are common
2. The fact that C++ namespaces allow us to differentiate between a function (in `stan::math`) and a variable (in `::`) meant we could also let users call that function too. So, we allow it. 

Essentially, [Stan is a Lisp2](https://www.nhplace.com/kent/Papers/Technical-Issues.html), at least when it comes to the standard library. Emulating this same behavior in user functions is possible, but would require extra work due to the fact that we don't namespace-qualify UDF calls.

File: test/integration/good/compiler-optimizations/dune
Comment (nhuurre) [inline test/integration/good/compiler-optimizations/dune:79]: Apparently I don't how these tests work. Why isn't this list just `(glob_files *.stan)`? And was missing `overloaded-fn.stan` a bug in the test?

Comment (WardBrian) [reply test/integration/good/compiler-optimizations/dune:79]: I think it was a historical thing, where some of the optimizations could cause the compiler to crash so we didn’t run all of them? Not completely sure, it predates me on the project. 

I wouldn’t be opposed to just making them a glob like the other tests are

Comment (WardBrian) [reply test/integration/good/compiler-optimizations/dune:79]: Looked through the blame, and it seems that at first there was only one file in this directory and the dune file was written without the glob, and that seems to have just persisted without much interrogation.

**Comment by WardBrian** [comment]
I'm going to hold off on a final review/approval until after the [language meeting](https://discourse.mc-stan.org/t/monthly-language-dev-meeting/24161/55) tomorrow in case anyone has some novel thoughts on `extern` functions which are worth considering before this change

**Comment by WardBrian** [comment]
During the language meeting we discussed the `extern` syntax. I'll be opening another issue here to discuss it more fully but long story short is it isn't a blocker for this. @bob-carpenter is very excited this change is finally being made

**Review by WardBrian — APPROVED** [review]
LGTM!

**Pull request merged**
